_episodes/01-why-bother-performance.md

```diff
@@ -1,16 +1,133 @@
+>
+>
+{: .callout}
+
+
+> ## Calculate CPU hours
+> 
+> In the following example, assume that you are utilising all the available core in a node. Calculate the CPU hours required to run the following job. Each node has 40 cores.
+> 
+> ```
+> #SBATCH --nodes = 2
+> #SBATCH --time = 05:00:00
+> ```
+> {: .bash}
+> 
+> > ## Solution
+> >
+> > 400 CPUh.
+> >
+> > 
+> > 
+>{: .solution}
+{: .challenge}
+
+## How can performance be enhanced?
+
+(FIXME) Mention pool of workers, tuning, OpenMP, MPI, GPU and libraries and will refer to them later
+
+## HPC layout and partitioning
+
+> ## Enhancing Performance
+>
+> Enhacing performance is one of the most important things in computation. Which of these are viable ways of enhancing performance? There may be more than one correct answer.
+>
+> 1. Utilising more cores
+> 2. Utilising more nodes
+> 3. Increasing the simulation walltime and CPUh
+> 4. Optimising the code
+> 5. Utilising MPI (Message Passing Interface)
+> 6. Utilising OpenMP (Open Multi-Processing)
+> 7. Using performance enhancing libraries
+> 8. Use GPUs instead of CPUs
+> 9. Having a computer with higher flops
+> 10. Use a lower level coding language
+> 11. Splitting code up into smaller segments
+> 
+> > ## Solution
+> > 1. Yes, the more cores you have, the more work can be distributed across those cores.
+> > 2. Yes, nodes contain multiple cores, the number of which depends on the node. 
+> > 3. No, increasing walltime and CPUh only increases the duration the code will run for. It does not improve the code performance.
+> > 4. Yes, making the code more efficient by cutting out unnecessary loops and bottlenecks can greatly increase performance.
+> > 5. Yes, MPI can enable you to split your code into multiple processes distributed over multiple cores, known as parallel programming.
+> > 6. Yes, like MPI this is also parallel programming, but deals with threads, by splitting a process into multiple threads, each thread using a single core.
+> > 7. Yes, that is their purpose. Different libraries run on different architectures however.
+> > 8. Yes, GPUs are better at handling multiple tasks, whereas a CPU is better running singular tasks quickly.
+> > 9. Yes, the faster the computer, the faster the code can run, however, this may not always be possible logistically.
+> > 10. Yes, lower level languages (C, C++) run faster than higher level languages (Python) due to there being less abstraction to the machine to handle.
+> > 11. No, if you have a simulation that needs to be run from start to completion, splitting the code into segments won't be of any benefit and will waste compute resources.
+> >
+> > As you can see, there are a lot of right answers, however some methods work better than others, and it can entirely depend on the problem you are trying to solve.
+> >
+>{: .solution}
+{: .challenge}
+
+If these terms are unfamiliar now, don't worry as we will cover them over the duration of this course.
+
```

Comment:
Episode 1/line 97/ MCQ: In stead of giving 11 choices, can we limit it to 4 ? E.g. number of  cores and number of nodes can be clubbed together for this question. Then we can have another MCQ asking the user how can we increase the number of cores? Here we can offer again 4 different choice like having a new laptop with more cores, using more nodes, using hyperthreading, etc.

Reply:
Changed to one of 4 and one of 5, how to enhance performance and how to increase core count.